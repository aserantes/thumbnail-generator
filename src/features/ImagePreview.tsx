import React from "react";
import {
  ButtonRow,
  ImageInfo,
  ImageInfoTextBlock,
  ImageInfoTextItem,
  ImagePreviewWrapper,
  Thumbnail,
} from "../components/Common";
import { ProcessButton, RestartButton } from "../components";
import { getName, getPath, getSize, getType } from "../store/selectors";
import { useSelector } from "react-redux";

export function ImagePreview() {
  const name = useSelector(getName);
  const type = useSelector(getType);
  const size = useSelector(getSize);
  const path = useSelector(getPath);
  const sizeText = `${(size / 1024).toFixed(2)} Kb`;

  return (
    <ImagePreviewWrapper data-testid="ImagePreview-wrapper">
      <ImageInfo>
        <Thumbnail alt="Image Preview" src={path} />
        <ImageInfoTextBlock>
          <ImageInfoTextItem>
            <strong>Name:&nbsp;</strong> {name}
          </ImageInfoTextItem>
          <ImageInfoTextItem>
            <strong>Type:&nbsp;</strong> {type}
          </ImageInfoTextItem>
          <ImageInfoTextItem>
            <strong>Size:&nbsp;</strong> {sizeText}
          </ImageInfoTextItem>
        </ImageInfoTextBlock>
      </ImageInfo>
      <ButtonRow>
        <RestartButton />
        <ProcessButton />
      </ButtonRow>
    </ImagePreviewWrapper>
  );
}
/*
const onSubmit = (e) => {
  e.preventDefault();
  const formData = new FormData();
  formData.append("file", file);

  // Pass a setUploadPercentage as callback here
  putData(formData, { onUploadProgress: setUploadPercentage )
    .then((response) => { 
      console.log(response);
    });
};
export function putData(formData, { onUploadProgress }) {
  return axios.put('/update', formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
    onUploadProgress : (progressEvent) => {
      const progress = parseInt(Math.round((progressEvent.loaded * 100) / progressEvent.total));
      // Update state here
      onUploadProgress(progress);
    },
  });
}

*/
