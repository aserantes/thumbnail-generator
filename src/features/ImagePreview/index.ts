export { ImagePreview } from "./ImagePreview";

export {
  initialState,
  imagePreviewReducer,
  changeTest,
} from "./ImagePreviewSlice";
