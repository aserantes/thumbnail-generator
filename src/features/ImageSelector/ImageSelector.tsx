import React from "react";
import styled from "@emotion/styled";

const Container = styled.div`
  background-color: orange;
  padding: 20px;
`;

export const ImageSelector = () => {
  return <Container>ImageSelector Component</Container>;
};
