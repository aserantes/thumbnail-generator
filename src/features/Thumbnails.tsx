import React from "react";
import thumb120 from "../thumbnails/thumb-120x120.png";
import thumb160 from "../thumbnails/thumb-160x120.png";
import thumb400 from "../thumbnails/thumb-400x300.png";
import { ImageInfo, Thumbnail2, ThumbnailsWrapper } from "../components/Common";

export function Thumbnails() {
  return (
    <ThumbnailsWrapper data-testid="Thumbnails-wrapper">
      <ImageInfo>
        <a download href={thumb120}>
          <Thumbnail2
            alt="Image Preview"
            height="120px"
            src={`/thumbnails/thumb-120x120.png?t=${Date.now()}"`}
            width="120px"
          />
        </a>
      </ImageInfo>
      <ImageInfo>
        <a download href={thumb160}>
          <Thumbnail2
            alt="Image Preview"
            height="120px"
            src={`/thumbnails/thumb-160x120.png?t=${Date.now()}"`}
            width="160px"
          />
        </a>
      </ImageInfo>
      <ImageInfo>
        <a download href={thumb400}>
          <Thumbnail2
            alt="Image Preview"
            height="300px"
            src={`/thumbnails/thumb-400x300.png?t=${Date.now()}"`}
            width="400px"
          />
        </a>
      </ImageInfo>
    </ThumbnailsWrapper>
  );
}
