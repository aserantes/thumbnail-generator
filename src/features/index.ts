export { ImagePreview } from "./ImagePreview";

export { ImageSelector } from "./ImageSelector";

export { Thumbnails } from "./Thumbnails";
