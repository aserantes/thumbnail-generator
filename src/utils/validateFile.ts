import {
  MAX_FILE_SIZE,
  MAX_FILE_SIZE_ERROR,
  VALID_FILE_TYPES,
  VALID_FILE_TYPES_ERROR,
} from "Constants";

interface FileInfo {
  size: number;
  type: string;
}

export const validateFile = (fileInfo: FileInfo) => {
  const { size, type } = fileInfo;
  let isValidType = false;
  const errors: string[] = [];

  VALID_FILE_TYPES.forEach((validType) => {
    if (!isValidType) {
      if (validType.includes(type)) {
        isValidType = true;
      }
    }
  });

  const lala = "lala".includes("la");

  console.log(lala);

  if (size > MAX_FILE_SIZE) errors.push(MAX_FILE_SIZE_ERROR);

  if (VALID_FILE_TYPES.includes(type)) errors.push(VALID_FILE_TYPES_ERROR);

  return errors;
};
