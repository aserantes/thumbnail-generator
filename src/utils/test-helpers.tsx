import { render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
// TODO import the initial state of all slices, not just this example:
import { initialState as imagePreviewInitialState } from "features/ImagePreview";
import { RootState } from "store";

export const rootInitialState = {
  imagePreview: imagePreviewInitialState,
};

const mockStore = configureStore([thunk]);

export const renderWithRedux = (
  ui: JSX.Element,
  initialState: RootState = rootInitialState
) => {
  const store = mockStore(initialState);
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
  };
};
