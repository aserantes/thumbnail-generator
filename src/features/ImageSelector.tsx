import React, { useEffect } from "react";
import {
  BrowseButton,
  ButtonRow,
  CameraButton,
  DropZone,
  ImageSelectorWrapper,
} from "components";
import { getFileInfo } from "utils/fileHelpers";
import { selectors } from "store";
import { setFileToUploadData } from "store/fileToUploadSlice";
import { useDispatch, useSelector } from "react-redux";
import { useFileTypeAnalizer } from "hooks/useFileTypeAnalizer";

export function ImageSelector() {
  const chunkPath = useSelector(selectors.getChunkPath);
  const [result] = useFileTypeAnalizer(chunkPath);
  const dispatch = useDispatch();
  const handleDrop = (event: React.DragEvent<HTMLDivElement>): void => {
    const file = event.dataTransfer.files && event.dataTransfer.files[0];
    const fileInfo = getFileInfo(file);

    fileInfo && dispatch(setFileToUploadData(fileInfo));
  };
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const file = event.target.files && event.target.files[0];
    const fileInfo = getFileInfo(file);

    fileInfo && dispatch(setFileToUploadData(fileInfo));
  };

  useEffect(() => {
    if (result.ext !== "") {
      dispatch(
        setFileToUploadData({
          type: result.ext,
        })
      );
    }
  }, [result, dispatch]);

  return (
    <ImageSelectorWrapper data-testid="ImageSelector-wrapper">
      <DropZone onFileDrop={handleDrop} />
      <ButtonRow>
        <BrowseButton onFileChange={handleChange} />
        <CameraButton onFileChange={handleChange} />
      </ButtonRow>
    </ImageSelectorWrapper>
  );
}
