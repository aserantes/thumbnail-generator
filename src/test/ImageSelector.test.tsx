import React from "react";
import { render, screen } from "utils/test-helpers";
import { ImageSelector } from "features/ImageSelector";

describe("<ImageSelector />", () => {
  it("should render", () => {
    render(<ImageSelector />);
    const component = screen.getByTestId("ImageSelector");

    expect(component).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    expect(render(<ImageSelector />).asFragment()).toMatchSnapshot();
  });

  it("should show validation error if selected file > 5Mb", () => {});

  it("should show validation error if selected file type is not png|jpg", () => {});

  it("should call stageFile with input.file info if a valid file is selected", () => {});
});

/*

**************************************************
Works OK BUT it's not usefull. Keep with paradigm of testing-library, just test as a user story
**************************************************
it("should match snapshot", () => {});

  test("should upload the file", () => {
    const file = new File(["hello"], "hello.png", { type: "image/png" });

    render(<input data-testid="element" type="file" />);

    const input = screen.getByTestId("element") as HTMLInputElement;

    user.upload(input, file);

    if (input.files) {
      expect(input.files[0]).toEqual(file);
    }

    expect(input.files).toHaveLength(1);
  });

*/
