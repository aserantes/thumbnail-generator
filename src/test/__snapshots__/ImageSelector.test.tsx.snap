// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageSelector /> should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="css-1w5jlbf"
    data-testid="ImageSelector-wrapper"
  >
    <input
      class="css-0"
      data-testid="file-input"
      type="file"
    />
  </section>
</DocumentFragment>
`;
